## we are using keras predefined dataset to just understand the concept

In [19]:
from keras.datasets import mnist

# loading the dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()


In [20]:
# let's print the shape of the dataset
print("X_train shape", X_train.shape)
print("y_train shape", y_train.shape)
print("X_test shape", X_test.shape)
print("y_test shape", y_test.shape)

X_train shape (60000, 28, 28)
y_train shape (60000,)
X_test shape (10000, 28, 28)
y_test shape (10000,)


## keras imports for the dataset and building our neural network

In [21]:
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Conv2D, MaxPool2D
from keras.utils import np_utils

# Flattening the images from the 28x28 pixels to 1D 787 pixels

In [22]:
X_train = X_train.reshape(60000, 784)
X_test = X_test.reshape(10000, 784)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')


# normalizing the data to help with the training

In [23]:
X_train /= 255
X_test /= 255

## one-hot encoding using keras' numpy-related utilities

In [24]:
n_classes = 10
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)

Shape before one-hot encoding:  (60000,)
Shape after one-hot encoding:  (60000, 10)


# building a linear stack of layers with the sequential model

In [25]:
model = Sequential()
# hidden layer
model.add(Dense(100, input_shape=(784,), activation='relu'))
# output layer
model.add(Dense(10, activation='softmax'))

In [26]:
# looking at the model summary
model.summary()
# compiling the sequential model
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')
# training the model for 10 epochs
model.fit(X_train, Y_train, batch_size=128, epochs=10, validation_data=(X_test, Y_test))

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 100)               78500     
_________________________________________________________________
dense_7 (Dense)              (None, 10)                1010      
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
469/469 [==============================] - 2s 4ms/step - loss: 0.3709 - accuracy: 0.8992 - val_loss: 0.2017 - val_accuracy: 0.9428
Epoch 2/10
469/469 [==============================] - 2s 4ms/step - loss: 0.1735 - accuracy: 0.9507 - val_loss: 0.1442 - val_accuracy: 0.9584
Epoch 3/10
469/469 [==============================] - 2s 4ms/step - loss: 0.1261 - accuracy: 0.9634 - val_loss: 0.1175 - val_accuracy: 0.9648
Epoch 4/10
469/469 [==============================] - 2s 4ms/step - loss: 0.0992 - acc

## Now we will use and CNN **model** 

In [9]:
# keras imports for the dataset and building our neural network
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Conv2D, MaxPool2D, Flatten
from keras.utils import np_utils

In [10]:
# to calculate accuracy
from sklearn.metrics import accuracy_score

# loading the dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# building the input vector from the 28x28 pixels
X_train = X_train.reshape(X_train.shape[0], 28, 28, 1)
X_test = X_test.reshape(X_test.shape[0], 28, 28, 1)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

# normalizing the data to help with the training
X_train /= 255
X_test /= 255

In [11]:
# one-hot encoding using keras' numpy-related utilities
n_classes = 10
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)

Shape before one-hot encoding:  (60000,)
Shape after one-hot encoding:  (60000, 10)


In [12]:
# building a linear stack of layers with the sequential model
model = Sequential()
# convolutional layer
model.add(Conv2D(25, kernel_size=(3,3), strides=(1,1), padding='valid', activation='relu', input_shape=(28,28,1)))
model.add(MaxPool2D(pool_size=(1,1)))
# flatten output of conv
model.add(Flatten())
# hidden layer
model.add(Dense(100, activation='relu'))
# output layer
model.add(Dense(10, activation='softmax'))


In [13]:
# compiling the sequential model
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')

In [14]:
# training the model for 10 epochs
model.fit(X_train, Y_train, batch_size=128, epochs=20, validation_data=(X_test, Y_test))

Epoch 1/10
469/469 [==============================] - 4s 9ms/step - loss: 0.1903 - accuracy: 0.9434 - val_loss: 0.0839 - val_accuracy: 0.9734
Epoch 2/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0567 - accuracy: 0.9835 - val_loss: 0.0633 - val_accuracy: 0.9791
Epoch 3/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0338 - accuracy: 0.9898 - val_loss: 0.0506 - val_accuracy: 0.9835
Epoch 4/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0198 - accuracy: 0.9942 - val_loss: 0.0533 - val_accuracy: 0.9827
Epoch 5/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0133 - accuracy: 0.9961 - val_loss: 0.0530 - val_accuracy: 0.9835
Epoch 6/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0085 - accuracy: 0.9977 - val_loss: 0.0531 - val_accuracy: 0.9850
Epoch 7/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0057 - accuracy: 0.9985 - val_loss: 0.0599 - val_accuracy: 0.9837
Epoch 

## after adding more than more convolution network

In [ ]:
# one-hot encoding using keras' numpy-related utilities
n_classes = 20
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)

In [15]:
# building a linear stack of layers with the sequential model
model = Sequential()
# convolutional layer
model.add(Conv2D(25, kernel_size=(3,3), strides=(1,1), padding='valid', activation='relu', input_shape=(28,28,1)))
model.add(MaxPool2D(pool_size=(1,1)))

#adding convolution network
model.add(Conv2D(25, kernel_size=(3,3), strides=(1,1), padding='valid', activation='relu', input_shape=(28,28,1)))
model.add(MaxPool2D(pool_size=(1,1)))

# flatten output of conv
model.add(Flatten())
# hidden layer
model.add(Dense(100, activation='relu'))
# output layer
model.add(Dense(10, activation='softmax'))

In [16]:
# compiling the sequential model
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')

In [18]:
# training the model for 20 epochs
model.fit(X_train, Y_train, batch_size=128, epochs= 20, validation_data=(X_test, Y_test))

Epoch 1/20
469/469 [==============================] - 5s 12ms/step - loss: 0.0041 - accuracy: 0.9985 - val_loss: 0.0531 - val_accuracy: 0.9878
Epoch 2/20
469/469 [==============================] - 5s 11ms/step - loss: 0.0022 - accuracy: 0.9993 - val_loss: 0.0526 - val_accuracy: 0.9889
Epoch 3/20
469/469 [==============================] - 5s 11ms/step - loss: 0.0039 - accuracy: 0.9988 - val_loss: 0.0606 - val_accuracy: 0.9857
Epoch 4/20
469/469 [==============================] - 5s 11ms/step - loss: 0.0058 - accuracy: 0.9980 - val_loss: 0.0573 - val_accuracy: 0.9868
Epoch 5/20
469/469 [==============================] - 5s 11ms/step - loss: 0.0025 - accuracy: 0.9991 - val_loss: 0.0452 - val_accuracy: 0.9895
Epoch 6/20
469/469 [==============================] - 5s 11ms/step - loss: 0.0038 - accuracy: 0.9988 - val_loss: 0.0595 - val_accuracy: 0.9880
Epoch 7/20
469/469 [==============================] - 5s 11ms/step - loss: 0.0027 - accuracy: 0.9991 - val_loss: 0.0554 - val_accuracy: 0.9887

what we conclude the while doing with one layer of convolution we are getting accuracy of approximately 98 % while adding more layer convolution we are increased our accuracy with some small amount 